runtime/interpreter/errors.go

```diff
@@ -846,3 +846,17 @@ func (InvalidHexLengthError) IsUserError() {}
 func (InvalidHexLengthError) Error() string {
 	return "hex string has non-even length"
 }
+
+// InvalidatedResourceReferenceError is reported when accessing a reference value
+// that is pointing to a moved or destroyed resource.
+type InvalidatedResourceReferenceError struct {
+	LocationRange
+}
+
+var _ errors.UserError = InvalidatedResourceReferenceError{}
+
+func (InvalidatedResourceReferenceError) IsUserError() {}
+
+func (e InvalidatedResourceReferenceError) Error() string {
+	return "referenced resource has been moved or destroyed after taking the reference"
```

Comment:
this message is a little vague, why not something like

```suggestion
	return "attempted to follow a reference to a resource that has been moved or destroyed"
```

Reply:
I have mixed feelings about this.
Didn't want the error message to sound too technical, considering many smart-contract engineers may not be familiar with language internals/terms (i.e: "following a reference"). But at the same time, this may not be the best wording to put it (not my first language 😄), so I'm open to suggestions

Reply:
What about
```suggestion
	return "null reference error: underlying resource has been moved or destroyed"
```

Reply:
There are no "null references" (or "null"s in general) in Cadence, I'd avoid mentioning it, as it would just confuse developers.

The current wording aligns with the phrasing of the rest of the error message. We can do a pass of improving error messages in the future, this shouldn't block getting a Stable Cadence preview release out